quinn-proto/src/connection/pacing.rs

```diff
@@ -0,0 +1,110 @@
+//! Pacing of packet transmissions.
+
+use std::{
+    convert::TryInto,
+    time::{Duration, Instant},
+};
+
+/// A simple token-bucket pacer. The bucket starts full and has an adjustable capacity. Once the
+/// bucket is empty, further transmission is blocked. The bucket refills at a rate determined by
+/// the congestion window and estimated round-trip time.
+pub struct Pacer {
+    capacity: u64,
+    tokens: u64,
+    prev: Instant,
+}
+
+impl Pacer {
+    /// Obtains a new [`Pacer`].
+    pub fn new(capacity: u64, now: Instant) -> Self {
+        Self {
+            capacity,
+            tokens: capacity,
+            prev: now,
+        }
+    }
+
+    /// Return how long we need to wait before sending a packet.
+    ///
+    /// If we can send a packet right away, this returns `None`. Otherwise, returns `Some(d)`,
+    /// where `d` is the time before this function should be called again.
+    ///
+    /// The 5/4 ratio used here comes from the suggestion that N = 1.25 in the draft IETF RFC for
+    /// QUIC.
+    pub fn delay(
+        &mut self,
+        smoothed_rtt: Duration,
+        mtu: u16,
+        window: u64,
+        now: Instant,
+    ) -> Option<Instant> {
+        debug_assert_ne!(
+            window, 0,
+            "zero-sized congestion control window is nonsense"
+        );
+        if let Some(time_elapsed) = now.checked_duration_since(self.prev) {
+            self.prev = now;
+            if smoothed_rtt.as_nanos() == 0 {
+                return None;
+            }
+            let elapsed = (time_elapsed.as_nanos() / smoothed_rtt.as_nanos())
+                .try_into()
+                .unwrap_or(u64::max_value());
+
+            let new_tokens = (window.saturating_mul(5) / 4).saturating_mul(elapsed);
+            self.tokens = self.tokens.saturating_add(new_tokens).min(self.capacity);
+        }
```

Comment:
Would it make sense to split this into two separate methods--one that queries the next permitted transmit time (and only takes `&self`), and one that updates the state to account for a transmit taking place? I'm concerned that the behavior of this as written might be sensitive to the rate at which it's called, which has no particular bounds in either direction.

Reply:
It probably does.  I will try and see what happens.